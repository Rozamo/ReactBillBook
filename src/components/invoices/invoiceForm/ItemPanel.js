import React, { Component } from "react";
import { PropTypes } from "prop-types";
import AddItemIcon from "../icons/AddItemIcon";

class ItemPanel extends Component {
	constructor(props) {
		super(props);
		this.state = {
			items: [],
			selectedItems: [],
			showDDL: false,
			tot: 0,
		};
	}
	componentDidMount() {
		fetch("https://rzp-training.herokuapp.com/team2/items")
			.then((res) => res.json())
			.then(
				(result) => {
					this.setState({
						items: result.items,
					});
				}
			);
	}

	handleAddItemClick = () => {
		this.setState({ showDDL: true });
	};
	handleSelectChange = (event, changeAmount) => {
    const {selectedItems,items} = this.state;
    let {tot}=this.state;
		const idx = event.target.value;
		const item = items[idx];
    tot = Number(tot + item.amount / 100);
    
		selectedItems.push({ item, quantity: 1, id: idx });
		this.setState(
			{
				showDDL: false,
				selectedItems,
				tot,
			},
			() => changeAmount(this.state.tot, this.state.selectedItems)
		);
	};
	handleQuantityInputChange = (event, changeAmount) => {
    const{selectedItems}=this.state;
		const item = selectedItems[event.target.id];
		let {tot} = this.state;
		tot -= Number(
			(item.item.amount / 100) * (item.quantity ? item.quantity : 0)
		);
		item.quantity = event.target.value;
		tot += Number(
			(item.item.amount / 100) * (item.quantity ? item.quantity : 0)
		);
		selectedItems[event.target.id] = item;
		this.setState(
			{
				selectedItems,
				tot,
			},
			() => {
				changeAmount(this.state.tot, this.state.selectedItems);
			}
		);
	};
	handleDeleteClick = (event, changeAmount) => {
		const {selectedItems} = this.state;
			let {tot} = this.state;
		tot -=
			(selectedItems[event.target.id].item.amount / 100) *
			selectedItems[event.target.id].quantity;
		selectedItems.splice(event.target.id, 1);
		this.setState({ selectedItems, tot }, () => {
			changeAmount(this.state.tot, this.state.selectedItems);
		});
	};
	renderSelect() {
		return (
			<div className="item-btn">
				<select
					className="item_ddl"
					defaultValue="Select Items"
					onChange={(event) => {
						this.handleSelectChange(event, this.props.changeAmount);
					}}
				>
					<option value="Select Items">Select Items</option>
					{this.state.items.map((item, index) => {
						if (
							this.state.selectedItems.findIndex((e) => e.item === item) ===
							-1
						)
							return (
								<option key={index} value={index}>
									{item.name}
								</option>
              );
              else return <></>
					})}
				</select>
			</div>
		);
	}
	renderAddItemButton() {
		return (
			<div className="item-btn">
				<button id="add-item" onClick={() => this.handleAddItemClick()}>
					<AddItemIcon />
					Add Item
				</button>
			</div>
		);
	}
	render() {
		return (
			<div className="inv-item-panel">
				<table className="item-tbl" id="inv-item-tbl">
					<tbody>
						{this.state.selectedItems.length ? (
							<tr>
								<th style={{ width: "40%" }}>Items</th>
								<th style={{ width: "5%" }}>Quantity</th>
								<th>Price</th>
								<th>Amount</th>
							</tr>
						) : (
							<tr />
						)}
						{this.state.selectedItems.map((item, index) => (
							<tr key={index}>
								<td>{item.item.name}</td>
								<td>
									<input
										id={index}
										style={{ width: "100px" }}
										type="text"
										defaultValue="1"
										value={item.quantity}
										onChange={(event) => {
											this.handleQuantityInputChange(
												event,
												this.props.changeAmount
											);
										}}
									 />
								</td>
								<td>{item.item.amount / 100}</td>
								<td>
									{(item.item.amount * item.quantity) / 100}
									<i
										id={index}
										className="fa fa-trash"
										onClick={(event) =>
											this.handleDeleteClick(event, this.props.changeAmount)
										}
									 />
								</td>
							</tr>
						))}
					</tbody>
				</table>
				{this.state.showDDL ? this.renderSelect() : this.renderAddItemButton()}
			</div>
		);
	}
}

ItemPanel.propTypes={
  changeAmount:PropTypes.func.isRequired
}

export default ItemPanel;
