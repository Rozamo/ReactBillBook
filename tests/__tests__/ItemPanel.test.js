import React from "react";
import ItemPanel from "../../src/components/invoices/invoiceForm/ItemPanel";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";

const fakeData = {
  entity: "collection",
  items: [
    {
      name: "Jmp",
      email: "J@rzp.com",
      age: "32",
      address: "123, Charming Avenue",
      amount: 1231,
    },
  ],
};

test("Should render ItemPanel", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => new Promise((resolve) => resolve(fakeData)),
    })
  );
  const fakeChangeAmount = jest.fn();

  const { container } = render(<ItemPanel changeAmount={fakeChangeAmount} />);

  expect(container.querySelector("select") === null).toBeTruthy();
  expect(container).toMatchSnapshot();

  const addBtn = container.querySelector("button");

  fireEvent.click(addBtn);

  await waitFor(() => container);

  expect(container.querySelector("select") === null).toBeFalsy();
  expect(container.innerHTML).toContain("Jmp");
  expect(container).toMatchSnapshot();

  global.fetch.mockClear();
  delete global.fetch;
});
