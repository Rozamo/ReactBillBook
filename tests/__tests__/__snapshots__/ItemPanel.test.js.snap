// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ItemPanel 1`] = `
<div>
  <div
    class="inv-item-panel"
  >
    <table
      class="item-tbl"
      id="inv-item-tbl"
    >
      <tbody>
        <tr />
      </tbody>
    </table>
    <div
      class="item-btn"
    >
      <button
        id="add-item"
      >
        <svg
          class="bi bi-basket2-fill"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.929 1.757a.5.5 0 1 0-.858-.514L2.217 6H.5a.5.5 0 0 0-.5.5v1a.5.5 0 0 0 .5.5h.623l1.844 6.456A.75.75 0 0 0 3.69 15h8.622a.75.75 0 0 0 .722-.544L14.877 8h.623a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1.717L10.93 1.243a.5.5 0 1 0-.858.514L12.617 6H3.383L5.93 1.757zM4 10a1 1 0 0 1 2 0v2a1 1 0 1 1-2 0v-2zm3 0a1 1 0 0 1 2 0v2a1 1 0 1 1-2 0v-2zm4-1a1 1 0 0 0-1 1v2a1 1 0 1 0 2 0v-2a1 1 0 0 0-1-1z"
            fill-rule="evenodd"
          />
        </svg>
        Add Item
      </button>
    </div>
  </div>
</div>
`;

exports[`Should render ItemPanel 2`] = `
<div>
  <div
    class="inv-item-panel"
  >
    <table
      class="item-tbl"
      id="inv-item-tbl"
    >
      <tbody>
        <tr />
      </tbody>
    </table>
    <div
      class="item-btn"
    >
      <select
        class="item_ddl"
      >
        <option
          selected=""
          value="Select Items"
        >
          Select Items
        </option>
        <option
          value="0"
        >
          Jmp
        </option>
      </select>
    </div>
  </div>
</div>
`;
